File: automated-testing/examples/hello-world/test_hello_world.py
import unittest
from unittest.mock import MagicMock
import asyncio
from examples.hello_world.hello_world import main  # Assuming your hello_world.py is in examples/hello_world/

class TestHelloWorldAgent(unittest.IsolatedAsyncioTestCase):

    async def asyncSetUp(self):
        # Mock the worldlink_client to avoid actual bus interaction
        self.mock_client = MagicMock()
        self.mock_client.connect = MagicMock(return_value=asyncio.Future())
        self.mock_client.connect().set_result(None)
        self.mock_client.register_agent = MagicMock(return_value=asyncio.Future())
        self.mock_client.register_agent().set_result(None)
        self.mock_client.publish_message = MagicMock(return_value=asyncio.Future())
        self.mock_client.publish_message().set_result(None)
        self.mock_client.disconnect = MagicMock(return_value=asyncio.Future())
        self.mock_client.disconnect().set_result(None)

        # Patch the Client class in your module with the mock
        from examples.hello_world import hello_world
        hello_world.Client = MagicMock(return_value=self.mock_client)

    async def test_agent_registers(self):
        await main()
        self.mock_client.register_agent.assert_called_once()
        args, _ = self.mock_client.register_agent.call_args
        self.assertEqual(args[0], "hello_world_python")
        self.assertEqual(args[1]["name"], "Hello World Agent (Python)")

    async def test_agent_publishes_greeting(self):
        await main()
        # Check if publish_message was called with the correct topic and message
        expected_topic = "greeting"
        expected_message = {"text": "Hello from Python!"}
        published = False
        for call in self.mock_client.publish_message.mock_calls:
            args, _ = call
            if args[0] == expected_topic and args[1] == expected_message:
                published = True
                break
        self.assertTrue(published, f"Message not published to '{expected_topic}' with content: {expected_message}")

if __name__ == '__main__':
    unittest.main()

File: automated-testing/examples/hello-world/test_hello_world.js
const assert = require('assert');
const { Client } = require('worldlink-client'); // Assuming worldlink-client is installed
const helloWorld = require('../../../examples/hello-world/hello_world.js'); // Adjust path if necessary

// Mock the Client class
jest.mock('worldlink-client', () => {
    const mockClient = {
        connect: jest.fn().mockResolvedValue(),
        registerAgent: jest.fn().mockResolvedValue(),
        publishMessage: jest.fn().mockResolvedValue(),
        disconnect: jest.fn().mockResolvedValue(),
    };
    return { Client: jest.fn(() => mockClient) };
});

describe('HelloWorldAgent (JavaScript)', () => {
    let client;

    beforeEach(() => {
        // Reset the mock before each test
        Client.mockClear();
        client = new Client();
    });

    test('agent registers', async () => {
        await helloWorld.main();
        expect(client.registerAgent).toHaveBeenCalledTimes(1);
        const [agentId, manifest] = client.registerAgent.mock.calls[0];
        expect(agentId).toBe('hello_world_js');
        expect(manifest.name).toBe('Hello World Agent (JavaScript)');
    });

    test('agent publishes greeting', async () => {
        await helloWorld.main();
        expect(client.publishMessage).toHaveBeenCalledTimes(1);
        const [topic, message, senderId] = client.publishMessage.mock.calls[0];
        expect(topic).toBe('greeting');
        expect(message).toEqual({ text: 'Hello from JavaScript!' });
        expect(senderId).toBe('hello_world_js');
    });
});